libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.cpp

```diff
@@ -0,0 +1,325 @@
+        }
+    }
+}
+
+// output_armed - sends commands to the motors
+void AP_MotorsMatrix_6DoF_Scripting::output_armed_stabilizing()
+{
+    uint8_t i;                          // general purpose counter
+    float   roll_thrust;                // roll thrust input value, +/- 1.0
+    float   pitch_thrust;               // pitch thrust input value, +/- 1.0
+    float   yaw_thrust;                 // yaw thrust input value, +/- 1.0
+    float   throttle_thrust;            // throttle thrust input value, 0.0 - 1.0
+    float   forward_thrust;             // forward thrust input value, +/- 1.0
+    float   right_thrust;               // right thrust input value, +/- 1.0
+
+    // note that the throttle, forwards and right inputs are not in bodyframe, they are in the frame of the 'normal' 4DoF copter were pretending to be
+
+    // apply voltage and air pressure compensation
+    const float compensation_gain = get_compensation_gain(); // compensation for battery voltage and altitude
+    roll_thrust = (_roll_in + _roll_in_ff) * compensation_gain;
+    pitch_thrust = (_pitch_in + _pitch_in_ff) * compensation_gain;
+    yaw_thrust = (_yaw_in + _yaw_in_ff) * compensation_gain;
+    throttle_thrust = get_throttle() * compensation_gain;
+
+    // scale horizontal thrust with throttle, this mimics a normal copter
+    // so we don't break the lean angle proportional acceleration assumption made by the position controller
+    forward_thrust = get_forward() * throttle_thrust;
+    right_thrust = get_lateral() * throttle_thrust;
+
+
+    // set throttle limit flags
+    if (throttle_thrust <= 0) {
+        throttle_thrust = 0;
+        // we cant thrust down, the vehicle can do it, but it would break a lot of assumptions further up the control stack
+        // 1G decent probably plenty anyway....
+        limit.throttle_lower = true;
+    }
+    if (throttle_thrust >= 1) {
+        throttle_thrust = 1;
+        limit.throttle_upper = true;
+    }
+
+    // rotate the thrust into bodyframe
+    Matrix3f rot;
+    Vector3f thrust_vec;
+    rot.from_euler312(_roll_offset, _pitch_offset, 0.0f);
+
+
+    /*
+        upwards thrust, independent of orentaiton
+    */
+    thrust_vec.x = 0.0f;
+    thrust_vec.y = 0.0f;
+    thrust_vec.z = throttle_thrust;
+    thrust_vec = rot * thrust_vec;
+    for (i = 0; i < AP_MOTORS_MAX_NUM_MOTORS; i++) {
+        if (motor_enabled[i]) {
+            _thrust_rpyt_out[i] =  thrust_vec.x * _forward_factor[i];
+            _thrust_rpyt_out[i] += thrust_vec.y * _right_factor[i];
+            _thrust_rpyt_out[i] += thrust_vec.z * _throttle_factor[i];
+
+            if (fabsf(_thrust_rpyt_out[i]) >= 1) {
+                // if we hit this the mixer is probably scaled incorrectly
+                limit.throttle_upper = true;
+            }
+            _thrust_rpyt_out[i] = constrain_float(_thrust_rpyt_out[i],-1.0f,1.0f);
+        }
+    }
+
+
+    /*
+        rotations
```

Comment:
when we say, "rotations" I guess this is the part of the motor mixer that causes the vehicle to rotate in roll, pitch yaw?

Reply:
exactly, I will improve the comment